typed/merge.go

```diff
@@ -407,12 +406,12 @@ func (w *mergingWalker) doUntyped(t schema.Untyped) (errs ValidationErrors) {
 		// Merge using the lhs's type on a shallow copy of the merge walker,
 		// to get rid of the merged output 
 		w2 := *w
-		w2.typeRef = schema.DeduceType(w.lhs)
+		w2.typeSpecifier.Type = schema.DeduceType(w.lhs)
```

Comment:
It feels a little "leaky" that we're always doing that. Can we return a `TypeSpecifier` from that method for example?